File: backend/app/models/review.py
from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import SQLModel, Field, Relationship


class Review(SQLModel, table=True):
    """Review model for service ratings and comments."""
    
    __tablename__ = "reviews"
    
    # Primary key
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    
    # Foreign keys
    service_id: UUID = Field(foreign_key="services.id", index=True)
    user_id: UUID = Field(foreign_key="users.id")
    merchant_id: UUID = Field(foreign_key="merchants.id", index=True)
    
    # Review content
    rating: int = Field(ge=1, le=5, description="Rating from 1 to 5 stars")
    comment: Optional[str] = Field(default=None, description="Review comment")
    
    # Status
    is_active: bool = Field(default=True, description="Review status")
    
    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Relationships
    service: "Service" = Relationship(back_populates="reviews")
    user: "User" = Relationship(back_populates="reviews")
    merchant: "Merchant" = Relationship(back_populates="reviews")


class InteractionType(str, Enum):
    """User interaction type enumeration."""
    VIEW = "view"
    LIKE = "like"
    SAVE = "save"
    SHARE = "share"


class UserInteraction(SQLModel, table=True):
    """User interaction model for tracking engagement."""
    
    __tablename__ = "user_interactions"
    
    # Primary key
    id: UUID = Field(default_factory=uuid4, primary_key=True)
    
    # Foreign keys
    user_id: UUID = Field(foreign_key="users.id", index=True)
    service_id: UUID = Field(foreign_key="services.id", index=True)
    
    # Interaction details
    interaction_type: InteractionType = Field(description="Type of interaction")
    
    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    # Relationships
    user: "User" = Relationship(back_populates="interactions")
    service: "Service" = Relationship(back_populates="interactions")
    
    # Composite indexes for performance
    __table_args__ = (
        # Index for finding user's interactions
        {"sqlite_autoincrement": True},
    )

File: mobile/lib/features/reviews/domain/entities/review.dart
import 'package:equatable/equatable.dart';

/// Review entity (domain layer)
class Review extends Equatable {
  final String id; // UUID as string
  final String serviceId; // 9-digit string
  final String userId; // 9-digit string
  final String merchantId; // UUID as string
  final int rating; // 1-5
  final String? comment;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final ReviewUser? user;
  final ReviewService? service;

  const Review({
    required this.id,
    required this.serviceId,
    required this.userId,
    required this.merchantId,
    required this.rating,
    this.comment,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    this.user,
    this.service,
  });

  @override
  List<Object?> get props => [
        id,
        serviceId,
        userId,
        merchantId,
        rating,
        comment,
        isActive,
        createdAt,
        updatedAt,
        user,
        service,
      ];
}

/// User information in review
class ReviewUser extends Equatable {
  final String id; // 9-digit string
  final String name;
  final String? avatarUrl;

  const ReviewUser({required this.id, required this.name, this.avatarUrl});

  @override
  List<Object?> get props => [id, name, avatarUrl];
}

/// Service information in review
class ReviewService extends Equatable {
  final String id; // 9-digit string
  final String name;

  const ReviewService({required this.id, required this.name});

  @override
  List<Object?> get props => [id, name];
}

/// Paginated review response
class PaginatedReviewResponse extends Equatable {
  final List<Review> reviews;
  final int total;
  final int page;
  final int limit;
  final bool hasMore;
  final int totalPages;

  const PaginatedReviewResponse({
    required this.reviews,
    required this.total,
    required this.page,
    required this.limit,
    required this.hasMore,
    required this.totalPages,
  });

  @override
  List<Object?> get props => [reviews, total, page, limit, hasMore, totalPages];
}
